### PR TITLE
Parse genomes

### LINKED ISSUE #87: Parse genomes
Also parse genome information

### PR DESCRIPTION
**Pull request by toinehartman**
This includes parsing genome names and which segments are in them. Genome id/name pairs are saved in cache. However, which segments are in which genome is not (yet) cached. Because of the large amount of changes, this should probably be merged first.

Closes #87.

### CODE REVIEW

File: src/main/java/programminglife/gui/controller/GraphController.java [hunk 1]
```diff
@@ -165,6 +166,7 @@ public GenomeGraph getGraph() {
      * Setter for the graph.
      * @param graph The graph
      */
+    @NotNull
```
Comment (YDutchy) [inline src/main/java/programminglife/gui/controller/GraphController.java]: Good annotation

File: src/main/java/programminglife/model/Genome.java [hunk 2]
```diff
@@ -22,15 +34,12 @@
...
         this.segments = new TreeMap<>();
 
         // TODO: verify (assert) that the order of segments is correct.
 
-        int coordinate = 1; // coordinates are 1 based.
-        for (Segment s : segments) {
-            this.segments.put(coordinate, s);
-            coordinate += s.getSequenceLength();
-        }
+        segments.forEach(this::addSegment);
```
Comment (YDutchy) [inline src/main/java/programminglife/model/Genome.java]: Good change instead of the for loop

File: src/main/java/programminglife/parser/Cache.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+    /**
+     * Remove a cache file.
+     * @param name the name of the file to remove
+     * @return true if the file was deleted
+     * @throws IOException when strange things happen
+     */
+    public static boolean removeDB(String name) throws IOException {
+        return Files.deleteIfExists(Paths.get(toDBFile(name)));
+    }
```
Comment (YDutchy) [inline src/main/java/programminglife/parser/Cache.java]: Just wondering, where did clearDB go to?

File: src/test/java/programminglife/model/GenomeTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        assertEquals("CAT", longGenome.getSegment(1).getSequence());
+        assertEquals("CAT", longGenome.getSegment(3).getSequence());
+        assertEquals("TAG", longGenome.getSegment(4).getSequence());
+        assertEquals("TAG", longGenome.getSegment(6).getSequence());
+        assertEquals("GG", longGenome.getSegment(7).getSequence());
+        assertEquals("GG", longGenome.getSegment(8).getSequence());
+        assertEquals("TATA", longGenome.getSegment(9).getSequence());
+        assertEquals("TATA", longGenome.getSegment(12).getSequence());
+    }
+}
```
Comment (YDutchy) [inline src/test/java/programminglife/model/GenomeTest.java]: Will these tests be expanded in the near future?

File: src/main/java/programminglife/model/Genome.java [hunk 5]
```diff
@@ -22,15 +34,12 @@
      */
     public Genome(String name, List<Segment> segments) {
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/Genome.java]: Javadoc for this method is not up to date.

File: src/test/java/programminglife/model/GenomeTest.java (hunk 4 above)
Comment (toinehartman) [inline src/test/java/programminglife/model/GenomeTest.java]: As this handles most edge cases and branch coverage is 100%, not in the near future.

File: src/main/java/programminglife/parser/Cache.java (hunk 3 above)
Comment (toinehartman) [inline src/main/java/programminglife/parser/Cache.java]: In my opinion, an existing cache file with no contents is never useful, only confusing. Therefore, clear = remove now.

File: src/main/java/programminglife/parser/GraphParser.java [hunk 6]
```diff
@@ -50,15 +45,17 @@ public GraphParser(File graphFile) throws IOException {
...
-            int secondsElapsed = (int) ((System.nanoTime() - this.startTime) / 1000000000.d);
+            int secondsElapsed = (int) ((System.nanoTime() - startTime) / 1000000000.d);
             System.out.printf("[%s] Parsing took %d seconds%n", Thread.currentThread().getName(), secondsElapsed);
             this.setChanged();
             this.notifyObservers(this.graph);
         } catch (Exception e) {
-            DataManager.rollback();
+            try {
+                this.getGraph().rollback();
+            } catch (IOException eio) { }
```
Comment (mvanmeerten) [inline src/main/java/programminglife/parser/GraphParser.java]: Shouldn't something happen when this goes wrong? For example, alert the user and restart the application?

File: src/main/java/programminglife/parser/Cache.java [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+    /**
+     * Remove a cache file.
+     * @param name the name of the file to remove
+     * @return true if the file was deleted
+     * @throws IOException when strange things happen
+     */
+    public static boolean removeDB(String name) throws IOException {
+        return Files.deleteIfExists(Paths.get(toDBFile(name)));
```
Comment (MeAmAnUsername) [inline src/main/java/programminglife/parser/Cache.java]: Check if currently open, if so close first (otherwise removing is not possible)

File: src/main/java/programminglife/parser/GraphParser.java [hunk 8]
```diff
@@ -108,14 +105,14 @@ protected synchronized void parse(boolean verbose) throws IOException, UnknownTy
...
                         break;
                     case 'H':
-                        System.out.println(line);
+                        this.parseHeader(line);
                         break;
                     default:
                         throw new UnknownTypeException(String.format("Unknown symbol '%c'", type));
                 }
 
                 if (Thread.currentThread().isInterrupted()) {
```
Comment (MeAmAnUsername) [inline src/main/java/programminglife/parser/GraphParser.java]: I don't think this does what you want, you should probably catch a InterruptedException, rollback, and then rethrow.

File: src/test/java/programminglife/gui/GuiControllerTest.java [hunk 9]
```diff
@@ -60,15 +57,15 @@ public void setUp() throws TimeoutException {
...
+        Cache.removeDB(TEST_File);
         FxToolkit.cleanupApplication(this.pl);
     }
 
     @AfterClass
     public static void tearDownClass() throws Exception {
-        DataManager.removeDB(TEST_DB);
-        DataManager.removeDB(TEST_File);
+        Cache.removeDB(TEST_DB);
+        Cache.removeDB(TEST_File);
```
Comment (MeAmAnUsername) [inline src/test/java/programminglife/gui/GuiControllerTest.java]: `removeDB` was already called in `tearDown()`.

File: src/test/java/programminglife/model/GenomeGraphTest.java [hunk 10]
```diff
@@ -33,24 +30,24 @@ public static void setUpClass() throws Exception {
...
     @After
     public void tearDown() throws Exception {
-        DataManager.clearDB(TEST_DB);
+        graph.removeCache();
     }
 
     @AfterClass
     public static void tearDownClass() throws Exception {
-        DataManager.removeDB(TEST_DB);
+        Cache.removeDB(TEST_DB);
```
Comment (MeAmAnUsername) [inline src/test/java/programminglife/model/GenomeGraphTest.java]: Doesn't this do the same as `graph.removeCache` in the `tearDown()` method?

File: src/test/java/programminglife/model/GenomeTest.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+        assertEquals(0, emptyGenome.length());
+        assertEquals("empty test genome", emptyGenome.getName());
+    }
+
+    @Test(expected = NoSuchElementException.class)
+    public void segmentOutOfBoundsTest() throws Exception {
+        // As this coordinate does not exist, an Exception should be thrown and the test should fail()
+        longGenome.getSegment(13).getSequence();
+        fail();
```
Comment (MeAmAnUsername) [inline src/test/java/programminglife/model/GenomeTest.java]: This is not needed, tests fail automatically when an expected exception is not thrown (or when they throw a different exception)

File: src/test/java/programminglife/parser/GraphParserTest.java [hunk 12]
```diff
@@ -117,8 +116,7 @@ public void update(Observable o, Object arg) {
                 GenomeGraph graph = (GenomeGraph) arg;
 
                 assertEquals(new File(TEST_PATH).getName(), graph.getID());
-                // TODO fix this test
-//                assertEquals("GTC", DataManager.getSequence(8));
+                assertEquals("GTC", graph.getSequence(8));
```
Comment (MeAmAnUsername) [inline src/test/java/programminglife/parser/GraphParserTest.java]: You fixed it, very good

File: src/main/java/programminglife/parser/Cache.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+     */
+    public Cache(String name) {
+        this.dbFileName = toDBFile(name);
+        System.out.printf("[%s] Setting up cache (%s)...%n", Thread.currentThread().getName(), this.dbFileName);
+        this.db = DBMaker.fileDB(new File(this.dbFileName))
+                .fileMmapEnableIfSupported()
+                .fileMmapPreclearDisable()
+                .cleanerHackEnable()
+                .closeOnJvmShutdown()
+                .checksumHeaderBypass()
```
Comment (toinehartman) [inline src/main/java/programminglife/parser/Cache.java]: https://jankotek.gitbooks.io/mapdb/content/db/

File: src/main/java/programminglife/parser/Cache.java [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+    /**
+     * Close the database.
+     * @throws IOException when unexpected things happen while closing
+     */
+    public void close() throws IOException {
+        if (!this.db.isClosed()) {
+            System.out.printf("[%s] Closing MapDB...%n", Thread.currentThread().getName());
+            this.db.close();
+        }
```
Comment (toinehartman) [inline src/main/java/programminglife/parser/Cache.java]: This is not relevant

File: src/main/java/programminglife/parser/Cache.java (hunk 7 above)
Comment (toinehartman) [inline src/main/java/programminglife/parser/Cache.java]: This is not possible from a static context. This method is only used to remove test files or remove an uninitialized method at startup.

File: src/test/java/programminglife/gui/GuiControllerTest.java (hunk 9 above)
Comment (toinehartman) [inline src/test/java/programminglife/gui/GuiControllerTest.java]: This is not really a problem, and to not interfere with other branches, this should be fixed globally for all tests in a separate PR.

File: src/test/java/programminglife/model/GenomeGraphTest.java (hunk 10 above)
Comment (toinehartman) [inline src/test/java/programminglife/model/GenomeGraphTest.java]: See above.
